lib/Sema/ConstraintSystem.cpp

```diff
@@ -1988,27 +2085,44 @@ ConstraintSystem::getTypeOfMemberReference(
     type = type->replaceSelfParameterType(baseObjTy);
   }
 
-  // When accessing protocol members with an existential base, replace
-  // the 'Self' type parameter with the existential type, since formally
-  // the access will operate on existentials and not type parameters.
-  if (!isDynamicResult &&
-      baseObjTy->isExistentialType() &&
-      outerDC->getSelfProtocolDecl()) {
-    auto selfTy = replacements[
-      cast<GenericTypeParamType>(outerDC->getSelfInterfaceType()
-                                 ->getCanonicalType())];
-    type = type.transform([&](Type t) -> Type {
-      if (t->is<TypeVariableType>())
-        if (t->isEqual(selfTy))
-          return baseObjTy;
-      if (auto *metatypeTy = t->getAs<MetatypeType>())
-        if (metatypeTy->getInstanceType()->isEqual(selfTy)) {
-          auto constraint = baseObjTy;
-          if (auto existential = baseObjTy->getAs<ExistentialType>())
-            constraint = existential->getConstraintType();
-          return ExistentialMetatypeType::get(constraint);
-        }
-      return t;
+  // Superficially, protocol members with an existential base are accessed
+  // directly on the existential, and not an opened archetype, and we may have
+  // to adjust the type of the reference (e.g. covariant 'Self' erasure) to
+  // support certain accesses.
+  if (!isStaticMemberRefOnProtocol && !isDynamicResult &&
+      baseObjTy->isExistentialType() && outerDC->getSelfProtocolDecl() &&
+      // If there are no type variables, there were no references to 'Self'.
+      type->hasTypeVariable()) {
+    const auto selfGP = cast<GenericTypeParamType>(
+        outerDC->getSelfInterfaceType()->getCanonicalType());
+
+    // First, temporarily bring back the 'Self' generic parameter.
+    type = type.transformRec([&](TypeBase *t) -> Optional<Type> {
+      // Don't recurse into children unless we have to.
+      if (!type->hasTypeVariable())
+        return Type(t);
+
+      if (isa<TypeVariableType>(t) && t->isEqual(replacements.lookup(selfGP)))
+        return selfGP;
+
+      // Recurse.
+      return None;
+    });
+
+    // Then, type-erase occurences of covariant 'Self'-rooted type parameters.
```

Comment:
Typo `occurences`. This three-step transform makes a ton of sense. `getTypeOfMemberReference` is getting *huge*, so at some point I'd like us to factor the transform out into a separate function.